### ISSUE TITLE
About attack generation process

### ISSUE THREAD

**Issue by THU-Yu-Shang** [MengmeiZ/RoHe#1]
Thanks for the opening source, I'm concerned about **the detailed process of generating the adversarial pertubation** (it is said in the paper to be generated by the FGSM-based attacking method), **could this part of codes be shared**? Thanks very much!

**Comment by MengmeiZ** [comment]
Hi, the detailed process can be found in our paper Algorithm 2: Adversarial edges generator. (http://shichuan.org/doc/132.pdf)

**Comment by THU-Yu-Shang** [comment]
Thank you, another question is how to get the second term $\dfrac {\partial{M^\Phi}} {\partial{M^{R_i}}}$ in equation (18) when there's an indicator function?

**Comment by THU-Yu-Shang** [comment]
Thanks for reply, another question is how to get the second term $\dfrac{\partial{M^\Phi}}{\partial{M^{R_i}}}$ in equation(18),  considering there's an indicator function when obtaining $M^\Phi$ as shown in equation(14).

**Comment by MengmeiZ** [comment]
Take metapath 'APCPA' as an example:
1.load the specified adjacency matrix of 'AP'-path and 'PC'-path
2. if R_i is 'AP', you can set the requires_grad property of 'AP' matrix as True
3. calculate 'APCPA' matrix based on 'AP' matrix and 'PC' matrix with "torch.matmul" and "torch.clamp", which are differentiable
4. cacluate the GCN loss of classification for target node (also differentialbe)
5. then use this loss to calculate the gradient of  'AP' matrix (torch.autograd.grad(loss, 'AP' matrix))
6. generate adversarial edges with gradient
